crowbar_framework/lib/crowbar/repository.rb

```diff
@@ -0,0 +1,159 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Repository
+    attr_reader :platform, :id, :registry
+
+    class << self
+      def load!
+        @config = YAML.load_file(Rails.root.join("config/repos.yml"))
+      end
+
+      def registry
+        load! unless defined? @config
+        @config
+      end
+
+      def where(options = {})
+        platform = options.fetch :platform, nil
+        repo = options.fetch :repo, nil
+        check_all_repos.select do |r|
+          if platform
+            if repo
+              r.platform == platform && r.registry["name"] == repo
+            else
+              r.platform == platform
+            end
+          else
+            r.registry["name"] == repo
+          end
+        end
+      end
+
+      def all_platforms
+        registry.keys
+      end
+
+      def check_all_repos
+        repochecks = []
+        all_platforms.each do |platform|
+          repositories(platform).each do |repo|
+            check = self.new(platform, repo)
+            if platform != Crowbar::Product.ses_platform && Crowbar::Product.is_ses?
+              #FIXME skip the repo check until SES is ported to SLE12-SP1
+              check.registry["required"] = "optional"
+            end
+            repochecks << check
+          end
+        end
+        repochecks
+      end
+
+      def admin_platform
+        NodeObject.admin_node.target_platform
+      end
+
+      def web_port
+        Proposal.where(barclamp: "provisioner").first.raw_attributes["web_port"]
+      end
+
+      def repositories(platform)
+        registry[platform]["repos"].keys
+      end
+
+      def admin_ip
+        NodeObject.admin_node.ip
+      end
+    end
+
+    def initialize(platform, repo)
+      @platform = platform
+      @id = repo
+      @registry = Repository.registry[@platform]["repos"][@id]
+      @url = url
+    end
+
+    def available?
+      all_repo_dirs.include?(@registry["name"]) && check_repo_tag
+    end
+
+    def valid_key_file?
+      if File.exist?(repomd_key_path)
+        md5 = Digest::MD5.hexdigest(File.read(repomd_key_path))
+        repomd_key_md5 == md5
+      else
+        false
+      end
+    end
+
+    def repodata_path
+      "/srv/tftpboot/#{@platform}/repos/#{@registry['name']}/repodata"
+    end
+
+    def repomd_key_md5
+      @registry["repomd"]["md5"]
+    end
+
+    def repomd_key_path
+      @registry["repomd"]["key"] || "#{repodata_path}/repomd.xml.key"
+    end
+
+    def url
+      @registry["url"] || \
+        "http://#{Repository.admin_ip}:#{Repository.web_port}/#{@platform}/repos/#{@registry['name']}"
+    end
+
+    def active?
+      !bag_item.nil?
+    end
+
+    def to_databag!
+      repository_item = Chef::DataBagItem.new
+      repository_item.data_bag "repositories"
+      repository_item["id"] = @id
```

Comment:
I don't think that's correct, since `@id` is per platform and multiple platforms can have a repo with the same id, but from a data bag perspective, you need a unique id.